### ISSUE TITLE
opensnoop exits with "oh no"

### ISSUE THREAD

**Issue by brayniac** [rust-bpf/rust-bcc#51]
In #48 we received a report that the opensnoop example wasn't working. The exact changes in that PR led me to close without merge. At the time, we didn't have tests in TravisCI for this example. It tested working on my machine. #50 attempts to add that example to CI and we see the issue there.

This issue is for trying to track down why this seems to work for some systems but not others and lead to a resolution. 

Systems known to be failing:
Fails with "oh no":
```
TravisCI VM infrastructure
OS Ubuntu Xenial
Kernel 4.15.0-1028-gcp x86_64
Rust 1.37.0
BCC 0.6.0->0.10.0
```

Fails with Segfault:
```
VMWare Fusion VM
OS Ubuntu Bionic
Kernel 4.15.0-58-generic x86_64
Rust 1.27.0
BCC 0.10.0
```

Systems shown to be working:
```
VMWare Fusion VM
OS Arch
Kernel 5.2.3-arch1-1-ARCH x86_64
Rust 1.36.0 / 1.37.0
BCC 0.10.0
```
```
VMWare Fusion VM
OS Debian Buster
Kernel 4.14.128 x86_64
Rust 1.37.0
BCC 0.10.0
```

**Comment by lilydjwg** [comment]
It fails here.

```
Arch Linux
Kernel 4.19.52 with uksm patch
rustc 1.38.0-nightly (311376d30 2019-07-18)
BCC 0.10.0.r87.89bb40a3-1
```

The `BPF_MAP_GET_NEXT_KEY` call fails at the 8th time. This is a snippet of strace log:

```
20721 openat(AT_FDCWD, "/sys/devices/system/cpu/online", O_RDONLY|O_CLOEXEC) = 12
20721 fcntl(12, F_GETFD)                = 0x1 (flags FD_CLOEXEC)
20721 read(12, "0-7\n", 32)             = 4
20721 read(12, "", 28)                  = 0
20721 close(12)                         = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 0, -1, PERF_FLAG_FD_CLOEXEC) = 12
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 12, 0) = 0x7fab4f4ee000
20721 ioctl(12, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 1, -1, PERF_FLAG_FD_CLOEXEC) = 13
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 13, 0) = 0x7fab4f4ad000
20721 ioctl(13, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 2, -1, PERF_FLAG_FD_CLOEXEC) = 14
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 14, 0) = 0x7fab4f46c000
20721 ioctl(14, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 3, -1, PERF_FLAG_FD_CLOEXEC) = 15
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 15, 0) = 0x7fab4f42b000
20721 ioctl(15, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 4, -1, PERF_FLAG_FD_CLOEXEC) = 16
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 16, 0) = 0x7fab4f3ea000
20721 ioctl(16, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 5, -1, PERF_FLAG_FD_CLOEXEC) = 17
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 17, 0) = 0x7fab4f3a9000
20721 ioctl(17, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 6, -1, PERF_FLAG_FD_CLOEXEC) = 18
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 18, 0) = 0x7fab4f368000
20721 ioctl(18, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = 0
20721 perf_event_open({type=PERF_TYPE_SOFTWARE, size=0 /* PERF_ATTR_SIZE_??? */, config=PERF_COUNT_SW_BPF_OUTPUT, ...}, -1, 7, -1, PERF_FLAG_FD_CLOEXEC) = 19
20721 mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_SHARED, 19, 0) = 0x7fab4f327000
20721 ioctl(19, PERF_EVENT_IOC_ENABLE, 0) = 0
20721 bpf(BPF_MAP_UPDATE_ELEM, {map_fd=6, key=0x55ce5cfa0750, value=0x55ce5cb8a7c0, flags=BPF_ANY}, 112) = 0
20721 bpf(BPF_MAP_GET_NEXT_KEY, {map_fd=6, key=0x55ce5cfa0750, next_key=0x55ce5cfa0750}, 112) = -1 ENOENT (没有那个文件或目录)
```

**Comment by brayniac** [comment]
@lilydjwg what platform? Virtual, baremetal? Any details there might help. Just having repro on travisci here right now. Which, is not great for iterating on this. Will be trying a VM based on Ubuntu Bionic next and hopefully a few baremetal systems. Might have to resume this on Monday though.

**Comment by lilydjwg** [comment]
It's my laptop without virtualization. It has four CPU physical cores and hyperthreading is enabled so there are eight logical cores, all online.

BTW, baremetal seems to have two meanings, either a piece of hardware without an OS or a dedicated physical server. Mine is closer to the latter.

**Comment by ethercflow** [comment]
@brayniac @lilydjwg I met this on ubuntu 18.04/5.2.11-050211-generic too. The hardware platform is a Dell server.

> Architecture:        x86_64
> CPU op-mode(s):      32-bit, 64-bit
> Byte Order:          Little Endian
> CPU(s):              40
> On-line CPU(s) list: 0-39
> Thread(s) per core:  2
> Core(s) per socket:  10
> Socket(s):           2
> NUMA node(s):        2
> Vendor ID:           GenuineIntel
> CPU family:          6
> Model:               79
> Model name:          Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz
> Stepping:            1
> CPU MHz:             1200.410
> CPU max MHz:         3100.0000
> CPU min MHz:         1200.0000
> BogoMIPS:            4399.98
> Virtualization:      VT-x
> L1d cache:           32K
> L1i cache:           32K
> L2 cache:            256K
> L3 cache:            25600K
> NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38
> NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39
> Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts md_clear flush_l1d

**Comment by ethercflow** [comment]
@brayniac @lilydjwg Hi, I think I've fix this problem, with PR https://github.com/rust-bpf/rust-bcc/pull/53. I found https://github.com/rust-bpf/rust-bcc/blob/master/src/perf.rs#L83 `bpf_get_next_key` the key to be set to be contiguous, even though one of the cpus in the middle is already offline. In this way https://github.com/rust-bpf/rust-bcc/blob/master/src/perf.rs#L81 (cpu, fd)'s cpu will be not currect. PTAL.

**Comment by brayniac** [comment]
This is resolved by #53
